Move task routes into a class based view

Bruce1347 opened the pull request:
This PR contains commits that move the function based views into a class based view.

It is motivated by the following points:

- These routes and their routing were written over 3 years ago, since then I learned more about REST design through work and the current naming is definitely not REST-y enough.
- The whole `views.py` is unorganized and messy, this is the occasion to regroup views that act on one kind of resource into the same class based view.

Bruce1347 committed:
refactor: move task deletion into a dedicated class based view

Bruce1347 committed:
refactor: move single task retrieval into a dedicated class based view

Bruce1347 committed:
refactor: move bits of task update into dedicated functions

Bruce1347 committed:
refactor: unify how tasks are retrieved for single tasks

This commit moves the task retrieval in a single function instead of
having each route replicate the same behavior.
It also fixes a pattern that should've been corrected a long time ago:
we do not want to rely on exceptions to check whether an object exists
in the database or not.

The new retrieval function is here to:
  * check if the provided ids are not null
  * check if the task actually exists in database

If one of the previous conditions aren't matched then the return value
is ``None`` and each view will return a 404.

Bruce1347 committed:
refactor: rewrite the update of a task into a class based view

Bruce1347 committed:
test: add a category factory

Bruce1347 committed:
refactor: use new PUT route for task update

This commit rewires the front to the newly introduced update route
(`POST /todo/lists/<list_id>/tasks/<task_id>`).

It also adds some fixes in the javascript files that were previously
uncaught.

Bruce1347 committed:
refactor(tests): remove `beta` prefix in test cases

Bruce1347 committed:
refactor: add a POST route for task creation

Bruce1347 committed:
chore(deps): add pydantic

Bruce1347 committed:
refactor: add Task schema for data validation

supakeen commented on yatdlm/todo/models.py:
```diff
@@ -194,6 +195,22 @@ def add_followup(self, followup, writer):
         followup.save()
 
+    def add_update_followup(self, writer, new_priority) -> None:
+        if new_priority is self.priority:
```
This should not be identity comparison between integers, use `==` for equality.

supakeen commented on yatdlm/todo/models.py:
```diff
@@ -250,6 +267,14 @@ def get_categories(self):
+
+        self.categories.clear()
+
+        self.categories.add(*categories_ids)
```
Seems to me this could 'just be' `self.categories = categories_ids` but I'm not familiar with Django's ORM.

supakeen commented on yatdlm/todo/models.py:
```diff
@@ -250,6 +267,14 @@ def get_categories(self):
 
         return [cat.as_dict() for cat in categories]
 
+    def set_categories(self, categories_ids: T.List[int]) -> None:
```
You keep using pre-3.9 syntax and 3.9 syntax interchangeably. Built-in `list` is subscriptable since that version so this could be `list[int]` like you do in other parts of the code (with `dict`).

supakeen commented on yatdlm/todo/models.py:
```diff
@@ -280,6 +305,26 @@ def as_dict(self, dates_format="d/m/Y"):
 
         return resp
 
+    def update(self, data: dict, logged_user: User) -> None:
```
Specify the `dict` type, even if it is `dict[str, T.Any]`.

supakeen commented on yatdlm/todo/models.py:
```diff
@@ -280,6 +305,26 @@ def as_dict(self, dates_format="d/m/Y"):
+
+        priority = data.get("priority")
+        title = data.get("title")
+        description = data.get("description")
```
Is it OK if these are `None`?

supakeen commented on yatdlm/todo/models.py:
```diff
@@ -280,6 +305,26 @@ def as_dict(self, dates_format="d/m/Y"):
+
+        if "categories" in data:
+            categories = [int(category_id) for category_id in data["categories"]]
+            self.set_categories(categories)
```
Personal preference would be to `self.set_categories(int(category_id) for ...)`.

supakeen commented on yatdlm/todo/schemas.py:
```diff
@@ -0,0 +1,49 @@
+
+
+class TaskSchema(BaseModel):
+    id: T.Union[int, None]
```
In modern enough Python you might prefer `id: int | None`.

supakeen commented on yatdlm/todo/schemas.py:
```diff
@@ -0,0 +1,49 @@
+    title: str
+    description: str = Field(alias="descr")
+    priority: int
+    categories: T.List[int]
```
Same as before, `list[int]` depending on Python version being targeted.

supakeen commented on yatdlm/todo/schemas.py:
```diff
@@ -0,0 +1,49 @@
+        # The input and output data key for that field is `descr`
+        # FIXME: replace `descr` by `description`
+        data["descr"] = obj.description
+
```
I really really dislike both the `T.Any` type here (it could just be `object` I think), the unspecified `dict`-types. And mutating the incoming `data` argument. Perhaps it'd be better to return a new `dict`.

supakeen commented on yatdlm/todo/schemas.py:
```diff
@@ -0,0 +1,49 @@
+    @classmethod
+    def from_orm(cls: "TaskSchema", obj: T.Any):
+        # The following fields cannot be copied "as is" and need a custom dumping method
+        custom_dump_fields = ["categories", "description"]
```
You can skip this list and use `custom_dump_fields_methods.keys()` where you *are* using it.

supakeen commented on yatdlm/todo/views.py:
```diff
@@ -539,3 +499,126 @@ def display_task_public(request, list_id, task_id):
+                    "errors": {
+                        "list_id": "Provided id is invalid",
+                        "task_id": "Provided id is invalid",
+                    }
```
Are they actually *both* invalid?

supakeen reviewed (commented):
Didn't spend a lot of time on this, just a quick scan.

Bruce1347 committed:
fix: prevent task creation with an invalid category id

Bruce1347 committed:
chore: Add lockfile related commands in Makefile

Bruce1347 committed:
chore(deps): Upgrade pydantic to 2.3.0

Bruce1347 committed:
refactor: add a propoer POST route for followup creation

Bruce1347 committed:
refactor: Add a GET route for tasks listing

Bruce1347 committed:
fix: Remove default owner_id to 1 for Tasks

Bruce1347 committed:
chore(deps): Add missed freezegun

Bruce1347 committed:
fix: Make id and task_no optional in TaskSchema

``id`` is an autogenerated and autoincremented field defined at the
first instantiation of an instance.

``task_no`` is a computed value that is incremented based on the number
of tasks in a given list.

These two fields are not needed at the creation of task, however the
previous implementation of the TaskSchema did not had these fields
marked as optional, making creation tests fail.

Bruce1347 committed:
feat: Use Django's ORM signals to set `task_no`

Bruce1347 committed:
feat(front): Use the new CRUD route for tasks

Bruce1347 committed:
chore: Add a pylint Makefile rule

Bruce1347 committed:
fix: Enable new Task route to create subtasks

Bruce1347 committed:
fix(deps): Re-add freezegun

Bruce1347 committed:
chore: Rewrite old style type hints

Bruce1347 committed:
feat: Handle rejection through PATCH

Bruce1347 committed:
feat: Use schemas for categories

Previous implmentation revolved around sending mono attributes dicts
with the category id.

However, since Categories are objects on their own, they need to be
handled as objects (i.e.: with proper routes and schemas).

This commit adds a CategorySchema and proper changes on the API and
front parts.

Bruce1347 committed:
fix: Include missed migration

This commit includes a missed migration on the task table which adds a
boolean column named ``rejected``.

Bruce1347 committed:
chore: add make rules

Bruce1347 committed:
feat: allow parent task filters on TaskListView

Bruce1347 committed:
fix: use ids instead of titles for render

This commit fixes the rendering of lists when a user is logged. The
previous version used the titles of lists without taking into account
that lists may share the same title.

Bruce1347 committed:
fix: make tasks no unique within lists

Bruce1347 committed:
fix: add parent list id in task schema

Bruce1347 committed:
feat: move from old tasks endpoint to new

Bruce1347 committed:
fix: rename ``orm_mode`` to ``from_attributes`` in schemas

Bruce1347 committed:
refactor: apply supakeen's suggestions

Bruce1347 committed:
fix: add missing ownership checks on list display

Bruce1347 committed:
fix(task): compare tasks count to a value instead

Bruce1347 committed:
refactor(factories): set a hashed password for users at creation

Bruce1347 committed:
test(categories): add missed tests

Bruce1347 committed:
fix: use a raw string for url regex matches

Bruce1347 committed:
chore(deps): add coverage

Bruce1347 committed:
feat: add a dedicated edition schema for tasks

Bruce1347 committed:
chore(deps): add ruff

Bruce1347 committed:
fix: return 404 on missing task

Bruce1347 committed:
chore(tests): add coverage make rule

Bruce1347 committed:
chore: exclude tests from coverage

Bruce1347 committed:
feat: add todolist creation list view

Bruce1347 committed:
chore: apply formatting on a bunch of files

Bruce1347 committed:
test: add more parametrized tests for rejection

Bruce1347 committed:
fix: finish migration from task.no to task.task_no

Bruce1347 committed:
test: add lists read route cases

Bruce1347 committed:
test: make tests faster, use MD5 hashing during testing

Bruce1347 committed:
refactor: make categories only deletable by their todolist owner

Bruce1347 committed:
fix: rename two missed tests

Bruce1347 committed:
chore: apply lint

Bruce1347 committed:
refactor(categories): use pydantic schemas for serialization and patch validation

Bruce1347 committed:
chore(ci): add github actions config

Bruce1347 committed:
fix(ci): set PYTHONPATH

Bruce1347 committed:
fix: add missed categories schemas

Bruce1347 committed:
chore(ci): bump actions to v5, add cache, use pipx

Bruce1347 committed:
chore(ci): test

Bruce1347 committed:
chore(ci): test
